### DISCUSSION TITLE
将RSS站点支持能力规则化，并独立于RSSHub作为规则库

### DISCUSSION THREAD

**Discussion by ttttmr** [DIYgod/RSSHub#12204] [Ideas]
tldr: 
将js代码转化为独立的规则，降低开发维护和社区贡献成本
rsshub与规则库，可以类比 adblock与过滤规则，或者[xray](https://github.com/chaitin/xray)与[xray poc](https://github.com/chaitin/xray/tree/master/pocs)

ref: https://blog.xlab.app/p/9c76cef/

最近Feed43凉了，替代品似乎也不多，为什么不能把RSSHub变成Feed43呢

RSSHub大部分的路由实现的套路都差不多

1. 从网页/API接口上抓取数据
2. 根据一定的规则，提取RSS需要的字段
3. 构建RSS

其中最核心部分有两个

1. 从哪里获取数据
2. 数据如何转换成需要的字段

核心部分都是通过js代码实现，想要支持一个新的RSS成本比较高

装环境，部署调试，提交pr，pr合并，更新自己的RSSHub，订阅新的RSS

有很多很多次，我想加一个RSS，都是因为这个功能Feed43也能实现，只要在网页上点几下就能制作好，而放弃RSSHub

这其实也不利于RSSHub的发展，扩充支持站点

## RSS-proxy

https://github.com/damoeb/rss-proxy

是一个比较理想的Feed43替代品

获取数据的地址和提取规则都是通过url传递，实时处理

增加新的RSS成本很低，但处理规则没有固化和分享

## 缝合

为什么不把RSSHub和RSS-proxy结合起来呢

将RSSHub处理路由和提取数据，参考RSS-proxy的模式，规则化

```
1. 输入地址
2. 链接转换成对应的API接口
3. 发起网络请求
4. 根据规则提取数据
5. 输出RSS
```

举一个例子

```
输入 https://a.com/users/ttttmr
转换成 https://a.com/api/users/news?user=ttttmr
发起网络请求，拿到json数据
提取数据
  - .items得到数据列表
    - .title得到标题
    - .link得到链接
然后构建RSS
```
结构化描述步骤2和4，就是一条规则

RSSHub引擎负责执行这个规则就好

而且2中规则还可以复用给[Radar](https://github.com/DIYgod/RSSHub-Radar)，来告知哪些可以订阅

这样新支持一个RSS会非常简单，把规则随便扔给一个实例测一下，没问题就可以直接用了

用户的实例订阅官方规则库，自动同步更新

- 关于规则如何定义和实现

可以参考XPath，CSS选择器，正则表达式，JSONPath等，或者[cel](https://github.com/google/cel-spec)

RSSHub自己定义实现也不是不行

- 关于路由实现

比如在同一个url下，可能存在多个api，也就是多个规则

可以规则定义一个名字来区分，来实现区分订阅比如`用户投稿`和`用户点赞`，类似

```
https://rsshub/feed?url=https://a.com/users/ttttmr&type=post
https://rsshub/feed?url=https://a.com/users/ttttmr&type=like
```

获取url参数，根据host过滤规则库，在规则库中遍历执行规则2，匹配出合适规则

根据type参数取出对应的规则，引擎执行，输出RSS

**Comment by ttttmr** [comment]
可能在有些场景下会有额外的问题
1. 维护站点登录/cookie，或者自动带Token请求与Token刷新
2. 多api合并制作一个rss

**Reply by valuex** [reply]
自部署或云端版的Check 酱已经实现了类似的功能，通过配置selector监控网页变化，然后生成rss。

**Reply by ttttmr** [reply]
这边好像没有pr关联更新，基于css selector做的基础rss转换已经实现了 https://github.com/DIYgod/RSSHub/pull/12882

文档可以看这个 https://docs.rsshub.app/zh/routes/other#transformation-html

也可以看这个 https://blog.xlab.app/p/1209a846/

未来还是希望能写个算法去自动识别元素，实现开箱即用的适配一些网站

**Reply by valuex** [reply]
这个太赞了，对于一些简单的网站非常好用！

**Reply by after9** [reply]
很赞的功能，要是url实现上能包装好看点就更好了

**Reply by ttttmr** [reply]
> 很赞的功能，要是url实现上能包装好看点就更好了

你说的是我网站里的工具吗，现在follow中支持了transform的功能，可以试试 https://github.com/RSSNext/Follow/pull/870

**Reply by after9** [reply]
> > 很赞的功能，要是url实现上能包装好看点就更好了
> 
> 你说的是我网站里的工具吗，现在follow中支持了transform的功能，可以试试 [RSSNext/Follow#870](https://github.com/RSSNext/Follow/pull/870)

我意思是能否做得到通过请求转发啥的，能让用户访问符合rsshub route规范的url 。而后台帮忙代理访问transform功能的地址 :)  ，这样对使用者感觉更友好点儿

**Reply by ttttmr** [reply]
> > > 很赞的功能，要是url实现上能包装好看点就更好了
> > 
> > 
> > 你说的是我网站里的工具吗，现在follow中支持了transform的功能，可以试试 [RSSNext/Follow#870](https://github.com/RSSNext/Follow/pull/870)
> 
> 我意思是能否做得到通过请求转发啥的，能让用户访问符合rsshub route规范的url 。而后台帮忙代理访问transform功能的地址 :) ，这样对使用者感觉更友好点儿

rsshub本身没有规则存储（未来应该也不会有），肯定不能简单传参

有转换规则托管的规则库话，可以只传url，详细规则从远程获取，链接就会精简一些
